### PR TITLE
Added configuration screen

### LINKED ISSUE #956: Config GUI
Apparently it is one of the most request features. Perhaps we could ask Configured dev if we could utilize their code as a starting point? It doesn’t need to be perfect first time around. Just something useable that we can improve over time

The config gui should show client and common configs. Server configs should only show if player has highest permission and then their changes can affect server config. Thoughts

### PR DESCRIPTION
**Pull request by HenryLoenwind**
It's not completely finished, but very close to a state I'd stop with.

This re-adds a generic configuration UI that can show and edit all the values the configuration system can handle. It is opt-in, and mods can use it as a base class to do their own thing. It mostly uses vanilla's configuration system with a couple of tweaks on top.

There are no breaking changes. It is mostly self-contained, with the exception of an addition to how lists are configured. For one, it allows mods to restrict the lengths of lists (in both directions), but more importantly, it gives the UI a way of creating new elements to add to lists.

I also fixed 2 configuration values that still had "forge" in the code but "neoforge" in the language file.

Open TODOs:

- [x] Warning screen after changing values that require a world/client restart (important)
- [x] "Reset to default" button
- [x] "Undo" button
- [x] More javadoc
- [x] PR to documentation repo
- [ ] Narrator for custom widget (help wanted as tagged in the source)
- Edit dedicated server config (out of scope)
- Discarded: Visually mark changed values (vs default/vs original value)

To use this in a mod, add `container.registerExtensionPoint(IConfigScreenFactory.class, ConfigurationScreen::new);` to your client-mod class constructor. (Although this probably works fine in the normal mod class, but the advanced constructor wouldn't.)

image
image
image
image
image
image

### CODE REVIEW

**Comment by HenryLoenwind** [comment]
Also, if someone would clean up neoforge's own configuration in regards to which ones are missing entries in the language file and which entries in the language file are zombies from old versions, that'd be great.

For this, note that the UI tries to get the tooltip from the language file and only uses the comment as fallback. While this doubles the text into two places, it also allows the tooltip to be translated and the comment to be present on servers.

**Comment by IchHabeHunger54** [comment]
Hello from the docs team!

The entire config article still needs a rework, so documentation for this system should best be added alongside that rework as a whole. In other words, don't let a missing docs PR stop you from merging this.

File: src/main/java/net/neoforged/neoforge/client/ClientNeoForgeMod.java [hunk 1]
```diff
@@ -24,11 +27,13 @@
 
 @Mod(value = "neoforge", dist = Dist.CLIENT)
 public class ClientNeoForgeMod {
```
Comment (TelepathicGrunt) [inline src/main/java/net/neoforged/neoforge/client/ClientNeoForgeMod.java:40]: Could you slip in an internal annotation for this class? It's technically an internal class that isn't for modders to call

Comment (HenryLoenwind) [reply src/main/java/net/neoforged/neoforge/client/ClientNeoForgeMod.java:40]: sure can do

**Comment by Technici4n** [comment]
This can now be rebased on top of the latest config changes. ;)

**Review by TelepathicGrunt — DISMISSED** [review]

**Comment by TelepathicGrunt** [comment]
I am approving this PR based off of end modder behavior and not the code portion. Just be sure to add to PR description that modders gets a console output of untranslated keys only if they visit the screens with the untranslated stuff and then exit the config gui. This should be stated so modders know this behavior exists.

Narrator support could be a future PR

**Comment by TelepathicGrunt** [comment]
@HenryLoenwind Looks like there's a conflicting file rn

**Review by TelepathicGrunt — APPROVED** [review]

File: src/main/java/net/neoforged/neoforge/client/gui/IConfigScreenFactory.java [hunk 2]
```diff
@@ -24,7 +23,7 @@ public interface IConfigScreenFactory extends IExtensionPoint {
     /**
      * Creates a new config screen. The {@code modListScreen} parameter can be used for a "back" button.
      */
-    Screen createScreen(Minecraft minecraft, Screen modListScreen);
+    Screen createScreen(ModContainer container, Screen modListScreen);
```
Comment (mezz) [inline src/main/java/net/neoforged/neoforge/client/gui/IConfigScreenFactory.java:26]: @HenryLoenwind This is a (small) breaking change, please include the breaking change tag on this commit for visibility or add a backward compatible method.

Comment (HenryLoenwind) [reply src/main/java/net/neoforged/neoforge/client/gui/IConfigScreenFactory.java:26]: @Matyrobbrt Your change ^

**Pull request merged**